src/NHibernate/Action/BulkOperationCleanupAction.cs

```diff
@@ -111,23 +111,30 @@ public void Execute()
 			// nothing to do
 		}
 
-		public BeforeTransactionCompletionProcessDelegate BeforeTransactionCompletionProcess
+		public IBeforeTransactionCompletionProcess BeforeTransactionCompletionProcess
 		{
-			get 
-			{ 
+			get
+			{
 				return null;
 			}
 		}
 
-		public AfterTransactionCompletionProcessDelegate AfterTransactionCompletionProcess
+		public IAfterTransactionCompletionProcess AfterTransactionCompletionProcess
 		{
 			get
 			{
-				return new AfterTransactionCompletionProcessDelegate((success) =>
-				{
-					this.EvictEntityRegions();
-					this.EvictCollectionRegions();
-				});
+				return new AfterTransactionCompletionProcess(
+					(success) =>
+					{
+						EvictEntityRegions();
+						EvictCollectionRegions();
```

Comment:
If the sync delegate is defined as a method, the Async Generator should then be able to generate its async counter-part, which could then be referenced as the async delegate here.

So from dev perspective, this is a three steps operation: define the sync method and use it as the sync delegate (along with a dummy async one or by using the sync only constructor), generate async, use the async counterpart as the async delegate.

Same for other places where there is those delegates to supply.

Reply:
It's in a backlog https://github.com/maca88/AsyncGenerator/issues/70

Reply:
> (along with a dummy async one or by using the sync only constructor)

Another way is to directly use the constructor that has both sync and async actions and use the async method even if it is not generated yet, this way you can save one code generation.

I've made [an example](https://github.com/maca88/nhibernate-core/commit/f17d3fc5b9b1fdb2968b8f01e3252b956d219f5e) for this PR, how the code could look like prior the generation of the async code. 

@hazzik For this PR, the mentioned generator feature wouldn't help, as here we need both sync and async methods to be passed in the constructor method, similar to what we have for [Future](https://github.com/nhibernate/nhibernate-core/blob/f3ccc55904b9d0ddbff6abaeb6e404bf377d1727/src/NHibernate/Impl/CriteriaImpl.cs#L409).